### PR TITLE
Adding support for ETCD encryption with KMS

### LINKED ISSUE #303: KMS etcd encryption
Leverage Key Management Service (KMS) plugin with  Azure Key Vault to encrypt etcd data.

https://docs.microsoft.com/en-us/azure/aks/use-kms-etcd-encryption

## Design notes

KMS etcd encryption doesn't work with System-Assigned Managed Identity

## Design decisions

Do we want a new, separate key vault for KMS etcd data?
Do we want to easily facilitate BYOK for KV?

### PR DESCRIPTION
**Pull request by Gordonby**
## PR Summary

Closes #303 

## TODO

- [x] Get working for public kv scenario
- [x] Add UI for KMS
- [x] Remove debug outputs from bicep
- [x] Delete KVPolicybased.bicep
~~- [ ] Create KMS Sample document that explains the bicep, that we can reference from the Azure Docs~~ (Lets pick up in a different PR thats more docs focussed)
- [x] Make sure it's not using the preview banner.
- [x] Show a warning message if privatelinks is enabled
- [x] Add an information message directing people to CSI KeyVault as a good alternative solution
- [x] Add byo capability
- [x] Test UI combinatons, evaulate if new messages needed
- [x] Test bicep byo kms

## UI

image

image

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] This PR is ready to merge and is not **Work in Progress**
- [x] Link to a filed issue
- [x] Screenshot of UI changes (if PR includes UI changes)

### CODE REVIEW

**Comment by Gordonby** [comment]
@khowling - Thinking of de-scoping the private link scenario here.
Reason being we need to create a private linked keyvault and then somehow get a key created, which becomes instantly harder because we've made it private and the deployment fails.

We could leverage a pattern of creating it public, and inserting the key then updating to be private, but this would be opaque and quite frankly a bit janky.

Thoughts?

**Comment by Gordonby** [comment]
`team meeting notes`

We're going to go with option 3, as the sort of customer that will care about KMS is the sort that will be using privatelinks anyway.

Look to support BYO KV and Key for private-link scenarios.

## New Parameter

```bicep
param existingKeyId string = ''
```

## UI

Only support BYO when privatelink=true and custom_vnet = true

Add a warning that the deploying user needs Owner RBAC on the KV to allow AKS to get to it.

**Comment by Gordonby** [comment]
@khowling - it's ready!

File: helper/src/components/portalnav.js [hunk 1]
```diff
@@ -330,6 +330,7 @@
   invalidFn('cluster', 'osDiskType', cluster.osDiskType === 'Ephemeral' && !VMs.find(i => i.key === cluster.vmSize).eph, 'The selected VM cache is not large enough to support Ephemeral. Select \'Managed\' or a VM with a larger cache')
   invalidFn('cluster', 'aad_tenant_id', cluster.enable_aad && cluster.use_alt_aad && cluster.aad_tenant_id.length !== 36, 'Enter Valid Directory ID')
   invalidFn('addons', 'registry', net.vnetprivateend && (addons.registry !== 'Premium' && addons.registry !== 'none'), 'Premium tier is required for Private Link, either select Premium, or disable Private Link')
+  invalidFn('cluster', 'keyVaultKmsByoKeyId', cluster.keyVaultKms === "byoprivate" && !cluster.keyVaultKmsByoKeyId.match('https:\/\/[^]+.vault.azure.net/keys/[^ ]+/[^ ]+$'), 'Enter valid KeyVault Versioned Key ID (https://YOURVAULTNAME.vault.azure.net/keys/YOURKEYNAME/KEYVERSIONSTRING)')
```
[the remark this answers is not in the document]
Comment (khowling) [reply helper/src/components/portalnav.js:333]: need to escape the . ??

File: bicep/main.bicep [hunk 2]
```diff
@@ -236,6 +237,125 @@ output keyVaultName string = keyVaultCreate ? kv.outputs.keyVaultName : ''
...
+  name: 'keyvaultKmsRbacs-${resourceName}'
+  params: {
+    keyVaultName: keyVaultKmsCreate ? kvKms.outputs.keyVaultName : ''
+    //We can't create a kms kv and key and do privatelink. Private Link is a BYO scenario
+    // rbacKvContributorSps : [
+    //   createAksUai && privateLinks ? aksUai.properties.principalId : ''
+    // ]
+    //This allows the Aks Cluster to access the key vault key
+    rbacCryptoUserSps: [
+      createAksUai ? aksUai.properties.principalId : ''
```
[the remark this answers is not in the document]
Comment (Gordonby) [reply bicep/main.bicep:291]: Yes, the module filters.

File: bicep/bicepconfig.json [hunk 3]
```diff
@@ -45,6 +45,9 @@
             },
             "use-stable-vm-image": {
                 "level": "warning"
+            },
+            "secure-secrets-in-params" : {
+                "level": "error"
```
[the remark this answers is not in the document]
Comment (Gordonby) [reply bicep/bicepconfig.json:50]: New bicep linter rule, am explicitly setting how we want to treat it

File: bicep/main.bicep [hunk 4]
```diff
@@ -38,7 +38,7 @@ param byoAKSSubnetId string = ''
 param byoAGWSubnetId string = ''
 
 //--- Custom, BYO networking and PrivateApiZones requires BYO AKS User Identity
-var createAksUai = custom_vnet || !empty(byoAKSSubnetId) || !empty(dnsApiPrivateZoneId)
+var createAksUai = custom_vnet || !empty(byoAKSSubnetId) || !empty(dnsApiPrivateZoneId) || keyVaultKmsCreateAndPrereqs || !empty(keyVaultKmsByoKeyId)
```
[the remark this answers is not in the document]
Comment (Gordonby) [reply bicep/main.bicep:41]: true dat.

if we're going to add any more conditions, then we should consider a better way.

**Review by khowling — APPROVED** [review]
its a good'n

**Pull request merged**